NIAD-266: Consume MESH part 2

grzegorz-robak opened the pull request.

grzegorz-robak committed:
NIAD-266: Consuming MESH messages

grzegorz-robak committed:
NIAD-266: Updated inbound UAT tests to use MESH client

grzegorz-robak committed:
NIAD-266: UAT test fix

grzegorz-robak committed:
NIAD-266: Added clearing of timestamp scheduler collection

grzegorz-robak committed:
NIAD-266: Jenkins file fix for test results

grzegorz-robak committed:
NIAD-266: Removed unnecessary file

grzegorz-robak committed:
NIAD-266: Renamed property

jaroslaw-wojciechowski commented on src/intTest/java/uk/nhs/digital/nhsconnect/nhais/uat/InboundMeshServiceUAT.java:
```diff
@@ -15,16 +20,39 @@
+    @BeforeEach
+    void setUp() throws Exception{
+        System.setProperty("NHAIS_SCHEDULER_ENABLED", "true"); //enable scheduling
+        while(!meshMailBoxScheduler.hasTimePassed(0)) { //First run creates collection in MongoDb
```
can't we use await() here?

jaroslaw-wojciechowski commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/MeshCypherDecoder.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import lombok.AllArgsConstructor;
```
unused imports

mattd-kainos commented on nhais-env-example.yaml:
```diff
@@ -19,8 +19,9 @@ NHAIS_MESH_ENDPOINT_PRIVATE_KEY: | #The content of the endpoint private key
   #keep 2 spaces indent for whole key
   -----END RSA PRIVATE KEY-----
-NHAIS_MESH_RECIPIENT_CODES: #recipient codes translation in format cypher=mesh_recipient_code (one per line)
+NHAIS_MESH_CYPHER_TO_MAILBOX: #recipient codes translation in format cypher=mesh_recipient_code (one per line)
```
Can you update the README as well?

grzegorz-robak replied:
Sure

mattd-kainos commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/model/mesh/MeshMessage.java:
```diff
@@ -19,4 +19,5 @@
      * The timestamp (ISO format, UTC) when this message was sent - used for debugging and tracing
      */
     private String messageSentTimestamp;
+    private String meshMessageId;
```
We don't need the message id for anything once it is published to the queue, but I think it is a good idea you've included it for traceability. We could even use this as the correlation id when distributed tracing is implemented 👍

mattd-kainos commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/MeshCypherDecoder.java:
```diff
@@ -0,0 +1,58 @@
+import uk.nhs.digital.nhsconnect.nhais.parse.EdifactParser;
+
+@Component
+public class MeshCypherDecoder {
```
This shouldn't be needed. These properties can be added to the TranslatedInterchange class.

mattd-kainos commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/service/EdifactToMeshMessageService.java:
```diff
@@ -22,6 +37,20 @@ public MeshMessage toMeshMessage(TranslatedInterchange translatedInterchange) {
+        //send to inbound queue
+        MeshMessage meshMessage = new MeshMessage();
+        meshMessage.setContent(edifactString);
+        meshMessage.setOdsCode(meshCypherDecoder.getSender(edifactString));
```
We have a TranslatedInterchange class that contains "other metadata needed for processing". These can be set when the interchange is encoded instead of needing parse the things we just contructed.

mattd-kainos commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/service/EdifactToMeshMessageService.java:
```diff
@@ -22,6 +37,20 @@ public MeshMessage toMeshMessage(TranslatedInterchange translatedInterchange) {
+        MeshMessage meshMessage = new MeshMessage();
+        meshMessage.setContent(edifactString);
+        meshMessage.setOdsCode(meshCypherDecoder.getSender(edifactString));
+        meshMessage.setWorkflowId(RecepMessage.isRecep(edifactString) ? WorkflowId.RECEP : WorkflowId.REGISTRATION);
```
Same here, TranslatedInterchange contains the interchangeType == workflowFlow id.

mattd-kainos commented on src/test/java/uk/nhs/digital/nhsconnect/nhais/mesh/MeshCypherDecoderTest.java:
```diff
@@ -8,9 +8,9 @@
 import uk.nhs.digital.nhsconnect.nhais.parse.EdifactParser;
 
-class MeshRecipientDecoderTest {
+class MeshCypherDecoderTest {
```
As above, should be able to remove this

grzegorz-robak committed:
NIAD-266: Reworked inbound UAT test scheduler initialization

mattd-kainos committed:
Added comments to MeshMessage to identify how properties should be used
Removed correlationId from MeshMessage - it is unused
Changed MeshClient signature to return MeshMessage

grzegorz-robak committed:
NIAD-266: Removed TranslatedInterchange as it was duplicating MeshMessage

grzegorz-robak committed:
NIAD-266: Removed isRecep methods as they are no longer needed

mattd-kainos commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/model/mesh/MeshMessage.java:
```diff
@@ -8,15 +8,39 @@
+    private String meshMessageId;
+
+    /**
+     * If SENDING TO MESH: Correlation id for GP system
```
"Value of OperationId response header for GP System" might describe it better

grzegorz-robak replied:
Agree

bartek-sarul commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/MeshWorkflowUnknownException.java:
```diff
@@ -0,0 +1,9 @@
+import uk.nhs.digital.nhsconnect.nhais.exceptions.NhaisBaseException;
+
+public class MeshWorkflowUnknownException extends NhaisBaseException {
+    public MeshWorkflowUnknownException(String message) {
```
`UnknownMeshWorkflowException` ?

grzegorz-robak replied:
Thrown when downloaded MESH message has different workflowId than we support (`NHAIS_REG` or `NHAIS_RECEP`)

bartek-sarul commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/model/mesh/MeshMessage.java:
```diff
@@ -8,15 +8,39 @@
-    private String odsCode;
+    /**
+     * If SENDING TO MESH: set to the HA Trading Partner Code (recipient)
+     * If DOWNLOADING FROM MESH: DO NOT USE
```
So maybe we should have `InboundMeshMessage` and `OutboundMeshMessage`?

grzegorz-robak replied:
Probably we should, maybe we could use interfaces for that, but I don't think it's that important right now

bartek-sarul commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/model/mesh/WorkflowId.java:
```diff
@@ -16,4 +19,11 @@
+
+    public static WorkflowId fromString(String workflowId){
+        return Arrays.stream(WorkflowId.values())
+                .filter(workflow -> workflow.workflowId.equalsIgnoreCase(workflowId))
```
Why not use exact matching and `Enum.valueof()` ?

grzegorz-robak replied:
`WorkflowId` enum has "human friendly" values, like `REGISTRATION`, but the actual value that I map here is `NHAIS_REG`

bartek-sarul commented on src/main/java/uk/nhs/digital/nhsconnect/nhais/service/InboundQueueService.java:
```diff
@@ -19,19 +20,26 @@
 @Slf4j
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
-public class InboundMeshService {
+public class InboundQueueService {
```
👍

bartek-sarul reviewed (changes requested):
Some comments

grzegorz-robak committed:
NIAD-266: Updated operationId javadoc in MeshMessage
